src/math/fft/fft.c

```diff
@@ -0,0 +1,180 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2020 Intel Corporation. All rights reserved.
+//
+// Author: Amery Song <chao.song@intel.com>
+//	   Keyon Jie <yang.jie@linux.intel.com>
+
+#include <sof/audio/coefficients/fft/twiddle.h>
+#include <sof/audio/buffer.h>
+#include <sof/audio/format.h>
+#include <sof/common.h>
+#include <sof/lib/alloc.h>
+#include <sof/math/fft.h>
+
+/*
+ * These helpers are optimized for FFT calculation only.
+ * e.g. _add/sub() assume the output won't be saturate so no check needed,
+ * and _mul() assumes Q1.31 * Q1.31 so the output will be shifted to be Q1.31.
+ */
+
+static void icomplex32_add(struct icomplex32 *in1, struct icomplex32 *in2, struct icomplex32 *out)
+{
+	out->real = in1->real + in2->real;
+	out->imag = in1->imag + in2->imag;
+}
+
+static void icomplex32_sub(struct icomplex32 *in1, struct icomplex32 *in2, struct icomplex32 *out)
+{
+	out->real = in1->real - in2->real;
+	out->imag = in1->imag - in2->imag;
+}
+
+static void icomplex32_mul(struct icomplex32 *in1, struct icomplex32 *in2, struct icomplex32 *out)
+{
+	out->real = ((int64_t)in1->real * in2->real - (int64_t)in1->imag * in2->imag) >> 31;
+	out->imag = ((int64_t)in1->real * in2->imag + (int64_t)in1->imag * in2->real) >> 31;
+}
+
+/* complex conjugate */
+static void icomplex_conj(struct icomplex32 *comp)
+{
+	comp->imag = SATP_INT32((int64_t)-1 * comp->imag);
+}
+
+/* shift a complex n bits, n > 0: left shift, n < 0: right shift */
+static void icomplex_shift(struct icomplex32 *input, int32_t n, struct icomplex32 *output)
+{
+	if (n > 0) {
+		/* need saturation handling */
+		output->real = SATP_INT32(SATM_INT32((int64_t)input->real << n));
+		output->imag = SATP_INT32(SATM_INT32((int64_t)input->imag << n));
+	} else {
+		output->real = input->real >> -n;
+		output->imag = input->imag >> -n;
+	}
+}
+
```

Comment:
looks like the above functions can be inline.

Reply:
> looks like the above functions can be inline.

yes, they can be inline, let me put them to header file also.